lib/compiler/src/beam_doc.erl

```diff
@@ -0,0 +1,561 @@
+    extract_documentation0(T, update_doc(State, {Doc, Anno})).
+
+
+%% NOTE: Terminal elements for the documentation, such as `-type`, `-opaque`, `-callback`,
+%%       and functions always need to reset the state when they finish, so that new
+%%       new AST items start with a clean slate.
+extract_documentation_from_funs([{function, Anno, F, A, [{clause, _, ClauseArgs, _, _}]}=_AST | T],
+                      #docs{exported_functions = ExpFuns}=State) ->
+    State1 = case (sets:is_element({F, A}, ExpFuns) orelse State#docs.export_all) of
+                true ->
+                   template_gen_doc({function, Anno, F, A, ClauseArgs}, State);
+                false ->
+                   State
+             end,
+   extract_documentation0(T, reset_state(State1));
+extract_documentation_from_funs([{function, Anno0, F, A, _Body}=_AST | T],
+                      #docs{doc = Doc0, exported_functions=ExpFuns}=State) ->
+    {Doc1, Anno1} = case Doc0 of
+                        none -> {none, Anno0};
+                        {Doc, Anno} -> {Doc, Anno}
+                    end,
+
+    State1 = case sets:is_element({F, A}, ExpFuns) orelse State#docs.export_all of
+               true ->
+                  {Slogan1, DocsWithoutSlogan} =
+                     %% First we check if there is a doc prototype
+                     case extract_slogan(Doc1, F, A) of
+                        undefined -> {io_lib:format("~p/~p",[F,A]), Doc1};
+                        SloganDocs -> SloganDocs
+                     end,
+                  Slogan = pick_slogan(Slogan1, State, F, A),
+                  AttrBody = {function, F, A},
+                  State0 = remove_exported_type_info(State),
+                  Result = gen_doc(Anno1, AttrBody, Slogan, DocsWithoutSlogan, State0),
+                  update_ast(function, State0, Result);
+               false ->
+                  State
+            end,
+   extract_documentation0(T, reset_state(State1)).
+
+pick_slogan(Slogan, #docs{slogan = none}, _F, _A) ->
+   Slogan;
+pick_slogan(_Slogan, #docs{slogan = {FunName, Vars, A}}, FunName, A) ->
+   VarString = create_variables(Vars),
+   unicode:characters_to_list(io_lib:format("~p(~s)", [FunName, VarString]));
+pick_slogan(Slogan, _, _F, _A) ->
+   Slogan.
+
+create_variables([]) ->
+   "";
+create_variables([Var | Vars]) ->
+   CommaSeparatedVars = create_variables(Vars, []),
+   atom_to_list(Var) ++ CommaSeparatedVars.
+
+create_variables([], Acc) ->
+   lists:reverse(Acc);
+create_variables([Var | Vars], Acc) ->
+   create_variables(Vars, atom_to_list(Var) ++ " ," ++ Acc).
+
+
+extract_documentation_from_cb([{attribute, Anno, callback, {{CB, A}, [Fun|_]}} | T], State) ->
+    Args = fun_to_varargs(Fun),
+    State1 = remove_exported_type_info(State),
+    State2 = template_gen_doc({callback, Anno, CB, A, Args}, State1),
+    extract_documentation0(T, reset_state(State2)).
```

Comment:
Types references from callbacks also need to be kept, just as `extract_spec_types` does for specs.